Add async test helper to timeout and provide a task_executor automatically

cecton opened the pull request:
# Use case

Since we introduced a helper to wait for blocks for our tests, it is now useful to also add a helper to put a time limit on our tests that run nodes..

See discussion https://github.com/paritytech/cumulus/pull/124#discussion_r452946110 for context.

See also some tests in polkadot: https://github.com/paritytech/polkadot/blob/master/node/test-service/tests/build-blocks.rs

# Solution

On your local machine (where you have control over the resources) it is preferable to have a timeout parameter that is small, but on the CI we don't know if the machine is loaded or not (tests may take more time than expected) therefore it is preferable for this value to be higher.

This solution resolves also some extra problems:
- async test: tests that are using tokio are fine but those using async-std at the moment might actually be flaky. Providing one opinionated way to do it will make sure that nobody ends up debugging bugs due to inconsistencies between the async runtimes.
- one common env var to rule them all (and in the darkness bind them)
- `task_executor` already defined for you

# To resolve

- [x] prevent task manager to be dropped before async runtime

seunlanlege commented on test-utils/derive/Cargo.toml:
```diff
@@ -0,0 +1,30 @@
+[package]
+name = "substrate-test-utils-derive"
```
name is wayy too long, maybe just test-utils? honestly could also live in a crate outside substrate.

cecton replied:
I named it that way for consistency with the other crates in that directory. Plus there is already a substrate-test-utils

seunlanlege replied:
yeah but it's not really substrate-specific is my point

seunlanlege replied:
I'm looking at the code and it looks like a `tokio-timed-test` to me :eyes:

cecton replied:
well... :sweat_smile: I guess!

The main reason I have this timeout is because in substrate you could start a node in a test and wait for blocks being produced. But I guess this could be one implementation for a more generic case which is "test timeout".

I did look for a crate that does exactly that but I couldn't find any. That is why I ended up doing it. Maybe I should make a PR to tokio instead.

cecton replied:
AH YES!!! I know why I didn't put it for tokio. There is a task executor that is proper to substrate

seunlanlege commented on test-utils/derive/src/lib.rs:
```diff
@@ -0,0 +1,101 @@
+			.into();
+			let timeout_task = tokio::time::delay_for(
+				std::time::Duration::from_secs(
+					option_env!("INTEGRATION_TEST_ALLOWED_TIME")
```
ooh ooh what about also taking the duration from the macro attributes?

a la:

```rust
#[test_utils::test(max_duration=60s)]
fn test_that_should_complete_under_60_seconds() {
}
```

cecton replied:
I thought about it but it is not really logical to do it. This is supposed to be a grace period for which a test can run on a machine, it's not supposed to vary between tests.

But... it DOES vary between different machines. Especially for the CI that uses a longer period of time (10min) while locally you could limit it to 1-3 min).

I feel like giving more choice will only encourage wrong behaviors: a test must not rely on time to run.

seunlanlege replied:
> a test must not rely on time to run.

ahh i see what you mean,

bkchr commented on test-utils/derive/src/lib.rs:
```diff
@@ -0,0 +1,101 @@
+			.into();
+			let timeout_task = tokio::time::delay_for(
+				std::time::Duration::from_secs(
+					option_env!("INTEGRATION_TEST_ALLOWED_TIME")
```
This requires me to recompile the test when I want to change this timeout..

cecton replied:
:bulb: you're right, will fix that!

cecton replied:
done

bkchr commented on test-utils/derive/src/lib.rs:
```diff
@@ -0,0 +1,102 @@
+		#header
+		#(#attrs)*
+		#vis #sig {
+			use futures::future::FutureExt;
```
This is a no go pattern in macros.

Macros should import required crates over the crate they are exported.

cecton replied:
I agree with you for macro_rules but this is a proc_macro crate, I can't re-export things like that:

```
error: `proc-macro` crate types currently cannot export any items other than functions tagged with `#[proc_macro]`, `#[proc_macro_derive]`, or `#[proc_macro_attribute]`
  --> test-utils/derive/src/lib.rs:21:1
   |
21 | pub use tokio;
   | ^^^^^^^^^^^^^^

error: `proc-macro` crate types currently cannot export any items other than functions tagged with `#[proc_macro]`, `#[proc_macro_derive]`, or `#[proc_macro_attribute]`
  --> test-utils/derive/src/lib.rs:22:1
   |
22 | pub use futures;
   | ^^^^^^^^^^^^^^^^
```

cecton replied:
I will investigate how this is usually done for proc-macros

cecton replied:
Ah yeah you said it but I misread. Ok I will re-export from test-utils

cecton replied:
done

bkchr commented on test-utils/derive/src/lib.rs:
```diff
@@ -0,0 +1,102 @@
+
+	if sig.inputs.len() != 1 {
+		let msg = "the test function accepts only one argument of type sc_service::TaskExecutor";
+		return Ok(syn::Error::new_spanned(&sig, msg).to_compile_error().into());
```
```suggestion
		return Err(syn::Error::new_spanned(&sig, msg));
```

bkchr commented on test-utils/derive/src/lib.rs:
```diff
@@ -0,0 +1,102 @@
+		_ => {
+			let msg =
+				"the test function accepts only one argument of type sc_service::TaskExecutor";
+			return Ok(syn::Error::new_spanned(&sig, msg).to_compile_error().into());
```
```suggestion
			return Err(syn::Error::new_spanned(&sig, msg));
```

bkchr commented on test-utils/derive/src/lib.rs:
```diff
@@ -0,0 +1,102 @@
+#[proc_macro_attribute]
+pub fn test(
+	args: proc_macro::TokenStream,
+	item: proc_macro::TokenStream,
```
You import it above.

cecton replied:
fixed

cecton commented:
@bkchr @seunlanlege I re-exported the macro in test-utils and made hidden imports for all the dependencies. I've put everything behind a feature flag ("macros").

I also changed the default timeout value to something more reasonable for a local run: 120. And I updated the CI config to use 600 instead.

cecton commented:
nah I changed my mind I put back 600 as default. It will work everywhere, that's better. And if someone needs fine tuning on a specific machine, they can use the env var.

bkchr commented on test-utils/derive/Cargo.toml:
```diff
@@ -0,0 +1,15 @@
+repository = "https://github.com/paritytech/substrate/"
+
+[dependencies]
+quote = "1"
```
Please put the exact version you are using.

cecton replied:
done

bkchr commented on test-utils/derive/Cargo.toml:
```diff
@@ -0,0 +1,15 @@
+
+[dependencies]
+quote = "1"
+syn = { version = "1", default-features = false }
```
Same

cecton replied:
done

bkchr commented on test-utils/derive/src/lib.rs:
```diff
@@ -0,0 +1,96 @@
+		#header
+		#(#attrs)*
+		#vis #sig {
+			use substrate_test_utils::futures::future::FutureExt;
```
This is still wrong. I said you should take a look at `runtime-interface`:

https://github.com/paritytech/substrate/blob/master/primitives/runtime-interface/proc-macro/src/utils.rs#L78

https://github.com/paritytech/substrate/blob/master/primitives/runtime-interface/proc-macro/src/utils.rs#L99

cecton replied:
done, see https://github.com/paritytech/substrate/pull/6651#discussion_r466273000

bkchr commented on test-utils/Cargo.toml:
```diff
@@ -6,6 +6,31 @@ edition = "2018"
 license = "Apache-2.0"
 homepage = "https://substrate.dev"
 repository = "https://github.com/paritytech/substrate/"
+autotests = false
```
Why?

cecton replied:
Because there are trybuild tests in the tests directory

bkchr replied:
Please just implement these tests as they are done everywhere else in the codebase.

bkchr commented on test-utils/Cargo.toml:
```diff
@@ -6,6 +6,31 @@ edition = "2018"
+trybuild = { version = "1.0", features = ["diff"] }
+
+[features]
+macros = [
```
Why do we need this?

cecton replied:
We don't "need" it. I can also remove the feature and the derive will be always included. That's what you want?

cecton replied:
done (I removed the feature flag)

cecton commented on test-utils/test-crate/Cargo.toml:
```diff
@@ -0,0 +1,16 @@
+targets = ["x86_64-unknown-linux-gnu"]
+
+[dev-dependencies]
+tokio = { version = "0.2.13", features = ["macros"] }
```
@bkchr I made this test-crate to check if it worked but I found out that tokio relies on the extern name tokio. The rest is fine and working though. I inspired my code from what you provided but I simplified a bit.

(I have not done the generate_crate_access yet!)

cecton replied:
I'm not sure what generate_crate_access is doing/solving. Do you have a small & quick explanation?

cecton commented:
I forgot to include the clean-up

cecton commented:
Nevermind. I wanted to make something to gracefully shutdown the task managers created during the test but I can't find a simple way to do that.

bkchr commented on test-utils/src/lib.rs:
```diff
@@ -17,6 +17,12 @@
 
+#[doc(hidden)]
+pub use futures;
+pub use substrate_test_utils_derive::test;
```
There are no docs how it works etc.

And pleas also mention that `tokio` is required in the `Cargo.toml` deps.

cecton replied:
I'm very disappointed because `#[cfg(test)]` seems to be false during doc tests... so I can write an example but I can't really test it or at least not automatically...

So I added a doc with a big warning about the tokio dependency but I can't add a concrete example with running a node since it would be unmaintained.

cecton commented:
Final review requested

bkchr commented on test-utils/src/lib.rs:
```diff
@@ -17,6 +17,27 @@
+/// Marks async function to be executed by an async runtime and provide a `TaskExecutor`, suitable
+/// to test environment.
+///
+///	# Example
```
```suggestion
/// # Example
```

bkchr commented on test-utils/src/lib.rs:
````diff
@@ -17,6 +17,27 @@
+///	# Example
+///
+///	```
+///	use tokio; // WARNING: you must have tokio in the dependency of your crate to use this macro!
````
```suggestion
```

This should be documented in the text above!

cecton replied:
done 0587d76

cecton replied:
fixed typo: e4b9ff1fc

bkchr commented on test-utils/src/lib.rs:
````diff
@@ -17,6 +17,27 @@
+///
+///	```
+///	use tokio; // WARNING: you must have tokio in the dependency of your crate to use this macro!
+///
````
```suggestion
```

bkchr commented on test-utils/src/lib.rs:
````diff
@@ -17,6 +17,29 @@
+///
+/// # Example
+///
+///	```
````
```suggestion
/// ```
```

bkchr commented:
bot merge
